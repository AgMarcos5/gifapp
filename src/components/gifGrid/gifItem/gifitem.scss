.gif_item{
    display: table-footer-group;
    position: relative;
    overflow: hidden;
    min-height: 200px;
    width: 100%;

    .hover_button{
        position: absolute;
        z-index: 1;
        height: 100%;
        width: 100%;
        display: flex;
        align-items: center;
        justify-content: center;
        background: linear-gradient(180deg, rgba(28,28,36,0) 16%, rgba(14,14,18,1) 100%);
        opacity: 0;
        transition: 1s;

        img{
            width: 40px;
            max-width: 40px;
            opacity: .9;
        }
    }

    &:hover {
        .hover_button{
            opacity: 1;
        }
        .gif_image{
            transform: scale(1.5);
        }
    }

    .gif_image{
        margin: 0 auto;
        position: absolute;
        height: 100%;
        min-width: 100%;
        transition: 1s;
    }
}