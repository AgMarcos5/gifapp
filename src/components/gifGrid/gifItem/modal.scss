@import 'src/styles/_colors';

.modal{
    width: 100vw;
    height: 100vh;    
    position: fixed;
    top: 0;
    left: 0;
    z-index: 2;

    .modal_background{
        -webkit-backdrop-filter: blur(5px);
        backdrop-filter: blur(5px);
        background: rgba(0,0,0,.5);
        width: 100%;
        height: 100%;
        position: absolute;
        display: flex;
        align-items: center;
        justify-content: center;
    }

    .modal_container{
        
        font-family: 'Roboto', sans-serif;
        max-width: min(500px,80vw);
        min-height: 100px;
        background: #0e0e12;
        position: relative;
        border-radius: 15px;
        box-shadow: rgb(5 5 6 / 20%) 0 7px 29px 0;
        overflow: hidden;

        .title{
            gap: 5px;
            display: flex;
            justify-content: space-between;
            align-items: center;
            padding: 0 10px;
            .title_info{
                display: flex;
                gap: 10px;
                align-items: center;
            }

            h3{
                font-weight: 300;
                text-transform: uppercase;
                font-size: 1em;
            }
        }

        .close_button{
            button{
                display: flex;
                align-items: center;
                margin: 0 auto;
                gap: 20px;
                border: 2px solid $gris;
                border-radius: 50px;
                background: transparent;
                color: $gris;
                padding: 5px;
                transition: .5s;
            }
            svg{
                fill: $grisClaro;
                width: 15px;
                height: 15px;
            }
            &:hover{
                transform: scale(1.05);
                button{
                    border-color: $baseColorHover;
                }
                svg{
                    fill: $grisClaro;
                }
            }
            &:active{
                transform: rotate(15deg);
            }
        }
        
        .fav_button{
            button{
                display: flex;
                align-items: center;
                margin: 0 auto;
                gap: 20px;
                border: 2px solid $gris;
                border-radius: 50px;
                background: transparent;
                color: $gris;
                padding: 5px;
                transition: .5s;
                svg{
                    fill: $grisClaro;
                    width: 15px;
                }

                &:hover{
                    transform: rotate(15deg) scale(1.05);
                    svg{
                        fill: $favColor;
                    }
                }
                &:active{
                    transform: rotate(15deg) scale(.6);
                }
            
                &.remove{
                    border-color: $favColor;
                    color: $favColor;
                    svg{
                        fill: $favColor2;
                    }
                }
            }
        }

        .modal_gif{
            display: flex;
            width: 100%;
        }
    }
}