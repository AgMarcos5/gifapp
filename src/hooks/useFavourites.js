import { useContext} from "react";
import { contextFavs } from "../context/favouriteContext";

const useFavourites = () => {
    const {
        state : {favourites},
        actions: {addFavourite,removeFavourite}
      } = useContext(contextFavs);

    return {favourites,addFavourite,removeFavourite}

}

export default useFavourites;