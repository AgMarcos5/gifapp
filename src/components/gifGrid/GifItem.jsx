import React from 'react'
import './gifitem.scss'

export const GifItem = ({url, title}) => {
  return (
    <div className='gif_item'>
     <img src={url} alt={title}/>
    </div>
  )
}
