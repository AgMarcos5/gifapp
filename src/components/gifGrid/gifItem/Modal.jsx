import React, { useEffect, useState } from 'react'
import useFavourites from '../../../hooks/useFavourites'
import './modal.scss'
import {ReactComponent as Star} from '../../../assets/star.svg'
import {ReactComponent as Close} from '../../../assets/close.svg'

import { motion, AnimatePresence } from "framer-motion"

export const Modal = ({id,title,url, modalState,setModalState}) => {

    const {favourites,addFavourite,removeFavourite} = useFavourites()
    const [isFav,setIsFav] = useState(false);


    useEffect( () => {
        if(favourites.find(fav => fav.id === id))
        {
            setIsFav(true)
        }
        else
        {
            setIsFav(false)
        }
    },[favourites])

    const data = {
        id,
        title,
        url
    }

    const variants = {
        visible: { opacity: 1 },
        hidden: { opacity: 0 },
      }
    
      const variantsContainer = {
        visible: { opacity: 1, scale: 1},
        hidden: { opacity: 0, scale: .1 },
      }
    
  return (
    <>
        {
            modalState &&
            <div className='modal'>
            
            <AnimatePresence>
            <motion.div
                key="modal"
                initial="hidden"
                animate="visible"
                variants={variants}
                exit={{ opacity: 0 }}
            >
                <div className='modal_background'>
                <motion.div
                    initial="hidden"
                    animate="visible"
                    transition={{ type: "spring", bounce: 0.25 }}
                    variants={variantsContainer}
                >
                    <div className='modal_container'>
                        <div className='title'>
                            <div className='title_info'>
                                <h3>{title}</h3>
                                <div className='fav_button'>
                                {
                                    isFav ?
                                    (<button onClick={() => removeFavourite(data)} className="remove"><Star/></button>)
                                    :
                                    (<button onClick={() => addFavourite(data)}><Star/></button>)
                                }
                                </div>
                            </div>
                            <div className='close_button'><button  onClick={()=>setModalState(null)}><Close/></button></div>
                           
                        </div>
                            <img className="modal_gif" src={url} alt={title}/>
                    </div>
                </motion.div>
                </div>
            </motion.div>
            </AnimatePresence>   
            </div> 
        }    
       
    </>
  )
}
