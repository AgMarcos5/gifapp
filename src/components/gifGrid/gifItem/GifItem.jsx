import React, { useState } from 'react'
import './gifitem.scss'
import open from '../../../assets/open.png'
import { Modal } from './Modal'

import PropTypes from 'prop-types'

export const GifItem = ({id,url, title}) => {

  const [modalState, setModalState] = useState(false)

  return (
    <>
    <div className='gif_item'>
      <div className='hover_button' onClick={() => setModalState(id)}>
          <img src={open} alt="open gif" />
      </div>
      <img className="gif_image" data-testid="gif" src={url} alt={title}/>
    </div>
    <Modal id={id} url={url} title={title} modalState={modalState} setModalState={setModalState}/>
    </>
  )
}

GifItem.propTypes = {
  id: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  title: PropTypes.string
}