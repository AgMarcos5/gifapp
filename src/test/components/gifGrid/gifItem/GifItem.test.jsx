

import { render, screen} from "@testing-library/react";
import { GifItem } from "../../../../components/gifGrid/gifItem/GifItem";
import useFavourites from "../../../../hooks/useFavourites";

jest.mock('../../../../hooks/useFavourites')

describe('Test en <GifItem/>', () => {
    
    const title = 'Dragon Ball';
    const url = 'https://dragon-ball.com/goku.jpg';
    const id = '123456'

    useFavourites.mockReturnValue({
        favourites: [],
        addFavourite: () => {},
        removeFavourite: () => {}
    })

    test('Debe coincidir con el snapshot ', () => {        
       const {container} = render( <GifItem title={title} url={url} id={id} /> );
       expect( container ).toMatchSnapshot();   
    });

    
    test('Debe mostrar URL y ALT', () => {
        render( <GifItem title={title} url={url} id={id} /> )
        //screen.debug();
        const {getByTestId} = screen;
        expect( getByTestId('gif').src ).toContain(url);
        expect( getByTestId('gif').alt ).toContain(title);  
    });

});