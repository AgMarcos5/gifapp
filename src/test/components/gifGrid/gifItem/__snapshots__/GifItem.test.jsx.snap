// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test en <GifItem/> Debe coincidir con el snapshot  1`] = `
<div>
  <div
    class="gif_item"
  >
    <div
      class="hover_button"
    >
      <img
        alt="open gif"
        src="/test-file-stub"
      />
    </div>
    <img
      alt="Dragon Ball"
      class="gif_image"
      data-testid="gif"
      src="https://dragon-ball.com/goku.jpg"
    />
  </div>
</div>
`;
